pkg/work/spoke/conditions/reader.go

```diff
@@ -0,0 +1,211 @@
+	}
+
+	return metav1.Condition{
+		Type:    rule.Condition,
+		Status:  status,
+		Message: message,
+		Reason:  reason,
+	}, err
+}
+
+func (s *ConditionReader) evaluateCelExpressions(obj *unstructured.Unstructured, expressions []workapiv1.CelConditionExpressions) (metav1.ConditionStatus, string, error) {
+	env, err := s.celEnv(cel.Variable("object", cel.DynType))
+	if err != nil {
+		return metav1.ConditionFalse, workapiv1.ConditionRuleInternalError, err
+	}
+
+	for _, expression := range expressions {
+		ast, iss := env.Compile(expression.Expression)
+		err = iss.Err()
+		if err != nil {
+			return metav1.ConditionFalse, workapiv1.ConditionRuleExpressionError, err
+		}
+
+		prg, err := env.Program(ast)
+		if err != nil {
+			return metav1.ConditionFalse, workapiv1.ConditionRuleInternalError, err
+		}
+
+		out, _, err := prg.Eval(map[string]any{
+			"object": obj.Object,
+		})
+		if err != nil {
+			return metav1.ConditionFalse, workapiv1.ConditionRuleExpressionError, err
+		}
+
+		switch result := out.Value().(type) {
+		case bool:
+			// If result is false return early
+			// Otherwise continue checking expressions
+			if !result {
+				return metav1.ConditionFalse, workapiv1.ConditionRuleEvaluated, nil
+			}
+		default:
+			err = fmt.Errorf("expected bool result, got %v", reflect.TypeOf(result))
+			return metav1.ConditionFalse, workapiv1.ConditionRuleExpressionError, err
+		}
+	}
+
+	return metav1.ConditionTrue, workapiv1.ConditionRuleEvaluated, nil
+}
+
+func (s *ConditionReader) getConditionMessageByRule(obj *unstructured.Unstructured, rule workapiv1.ConditionRule, result bool) (string, error) {
+	if rule.MessageExpression != "" {
+		env, err := s.celEnv(cel.Variable("object", cel.DynType), cel.Variable("result", cel.BoolType))
+		if err != nil {
+			return "", err
+		}
+
+		ast, iss := env.Compile(rule.MessageExpression)
+		err = iss.Err()
+		if err != nil {
+			// Trim CEL code snippets out of message
+			message := strings.Split(iss.String(), "\n | ")[0]
+			return message, err
+		}
+
+		prg, err := env.Program(ast)
+		if err != nil {
+			return "", err
+		}
+
+		out, _, err := prg.Eval(map[string]any{
+			"object": obj.Object,
+			"result": result,
+		})
+		if err != nil {
+			return "", err
+		}
+
+		switch message := out.Value().(type) {
+		case string:
+			if message != "" {
+				return message, nil
+			}
+		default:
+			return "", fmt.Errorf("expected message expression to have a string result, got %v", reflect.TypeOf(message))
+		}
+	}
+
+	if rule.Message != "" {
+		return rule.Message, nil
+	}
+	if result {
+		return "Manifest is " + rule.Condition, nil
+	}
+	return "Manifest is not " + rule.Condition, nil
+}
+
+func (s *ConditionReader) celEnv(opts ...cel.EnvOption) (*cel.Env, error) {
+	opts = append(
+		opts,
+		cel.OptionalTypes(),
+		ext.Strings(),
+		library.Lists(),
+		library.Regex(),
+		library.URLs(),
+		library.Quantity(),
+		library.IP(),
+		library.CIDR(),
+		library.Format(),
+		// TODO
+		// library.SemverLib(),
```

Comment:
@haoqing0110 Copied these over from your [CEL selector PR](https://github.com/open-cluster-management-io/ocm/pull/693). Might be a good idea to have a shared helper that establishes a common set of cel envs used by all OCM components, and then add extra envs specific to the use case on top of that

Reply:
This is a draft PR for some cel library and common things https://github.com/open-cluster-management-io/sdk-go/pull/105 .

Reply:
perfect!